### PR TITLE
Arrays/Record passing scheme refactoring

### PR DESCRIPTION
**Pull request by michaelvoronov**

### CODE REVIEW

File: crates/wit/tests/generation_tests/exports/arrays/expanded.rs [hunk 1]
```diff
@@ -55,7 +48,7 @@ pub unsafe fn __fce_generated_wrapper_func_inner_arrays_1(arg_0: u32, arg_1: u32
             }
             result
         }
-        let mut arg: Vec<u64> = Vec::from_raw_parts(offset as _, size as _, size as _);
+        let mut arg: Vec<u32> = Vec::from_raw_parts(offset as _, (2 * size) as _, (2 * size) as _);
```
Comment (boneyard93501) [inline crates/wit/tests/generation_tests/exports/arrays/expanded.rs:51]: how do you feel about making them (2,4) constants?

File: fluence/tests/compilation_tests/import_functions/ref_arrays.rs [hunk 2]
```diff
@@ -42,6 +42,8 @@ extern "C" {
 
     pub fn string_type(arg: &Vec<String>) -> Vec<String>;
 
+    pub fn bool_type(arg: &Vec<bool>) -> Vec<bool>;
```
Comment (boneyard93501) [inline fluence/tests/compilation_tests/import_functions/ref_arrays.rs:45]: i may be imagining things but i was under the impression that Vec<bool> is a no go.

File: fluence/src/call_parameters.rs [hunk 3]
```diff
@@ -16,12 +16,12 @@
 
 use fluence_sdk_macro::fce;
 
-use serde::Serialize;
-use serde::Deserialize;
+// use serde::Serialize;
+// use serde::Deserialize;
```
Comment (boneyard93501) [inline fluence/src/call_parameters.rs]: ?

File: crates/wit/src/parsed_type/vector_utils.rs [hunk 4]
```diff
@@ -155,7 +149,7 @@ pub(crate) fn generate_vector_deserializer(
     let values_deserializer = match value_ty {
         ParsedType::Boolean(_) => {
             quote! {
-                unimplemented!("Vector of booleans is not supported")
+                unimplemented!("Vector of booleans is unsupported")
```
Comment (boneyard93501) [inline crates/wit/src/parsed_type/vector_utils.rs]: i think not supported is a better message than  unsupported. tomatoe, tomato.

**Review by boneyard93501 — APPROVED** [review]

File: crates/wit/src/token_stream_generator/record_generator/field_values_builder.rs [hunk 5]
```diff
@@ -219,13 +235,31 @@ impl FieldValuesBuilder {
...
+                    raw_record[#value_id + 5],
+                    raw_record[#value_id + 6],
+                    raw_record[#value_id + 7],
+                ]);
+
+                String::from_raw_parts(offset as _, size as _, size as _)
+            };
+        };
+
+        self.value_id += 2 * 4;
```
[the remark this answers is not in the document]
Comment (michaelvoronov) [reply crates/wit/src/token_stream_generator/record_generator/field_values_builder.rs:260]: What about `2 * std::mem::size_of::<u32>()`?

File: crates/wit/src/export_ast_types.rs [hunk 6]
```diff
@@ -121,11 +130,25 @@ impl From<AstExternModItem> for ExternModItem {
...
+
+impl From<AstRecordFields> for RecordFields {
+    fn from(ast_record_item: AstRecordFields) -> Self {
+        match ast_record_item {
+            AstRecordFields::Named(fields) => {
+                let fields = fields.into_iter().map(Into::into).collect();
+                Self::Named(fields)
+            }
+            AstRecordFields::Unnamed(fields) => {
+                let fields = fields.into_iter().map(Into::into).collect();
```
[the remark this answers is not in the document]
Comment (michaelvoronov) [reply crates/wit/src/export_ast_types.rs:148]: Hmm, I don't how without more code lines

File: crates/wit/tests/generation_tests/exports/arrays/expanded.rs (hunk 1 above)
Comment (michaelvoronov) [reply crates/wit/tests/generation_tests/exports/arrays/expanded.rs:51]: It's an autogenerated code, but anyway I added consts.

File: fluence/tests/compilation_tests/import_functions/ref_arrays.rs (hunk 2 above)
Comment (michaelvoronov) [reply fluence/tests/compilation_tests/import_functions/ref_arrays.rs:45]: What do you mean? Please note that this's a test.

File: crates/wit/src/parsed_type/vector_utils.rs (hunk 4 above)
Comment (michaelvoronov) [reply crates/wit/src/parsed_type/vector_utils.rs]: Hmm, it's strange this file from an older version of sdk, now there is no such message at all.

File: crates/wit/src/token_stream_generator/record_generator/field_values_builder.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (boneyard93501) [reply crates/wit/src/token_stream_generator/record_generator/field_values_builder.rs]: sure .. although in the other files/occasions you just used 8 :)

[the remark this answers is not in the document]
Comment (michaelvoronov) [reply crates/wit/src/token_stream_generator/record_generator/field_values_builder.rs]: Re 8: it's used, beucase on that situations we considered types of 8 bytes width (u64/i64/...).

File: crates/main/src/export_allocator.rs [hunk 7]
```diff
@@ -14,23 +14,52 @@
...
+    log(format!(
+        "sdk.allocate: {} {} -> {}\n",
+        elem_count, elem_ty, allocated_mem
+    ));
 
-    log(format!("sdk.allocate: {:?}\n", size));
+    allocated_mem
+}
+
+macro_rules! alloc {
```
[the remark this answers is not in the document]
Comment (folex) [reply crates/main/src/export_allocator.rs:39]:
```rust
fn alloc<T>(count: usize) -> usize {
    let vec = Vec::<T>::with_capacity(count);
    let offset = vec.as_ptr() as usize;
    std::mem::forget(vec);
    offset
}
```

[the remark this answers is not in the document]
Comment (michaelvoronov) [reply crates/main/src/export_allocator.rs:39]: Thanks, good point!

File: crates/wit/src/utils.rs [hunk 8]
```diff
@@ -26,7 +26,7 @@ macro_rules! new_ident {
 macro_rules! prepare_global_data {
     ($fce_type: ident, $self: ident, $name: expr, $data: ident, $data_size: ident, $global_static_name: ident, $section_name: ident) => {
```
[the remark this answers is not in the document]
Comment (michaelvoronov) [reply crates/wit/src/utils.rs:27]: It's due the macro hygiene, this "trick" allows using them outside the macro.

File: crates/main/src/lib.rs [hunk 9]
```diff
@@ -57,14 +57,12 @@ pub use result::add_object_to_release;
 pub use module_manifest::MANIFEST_SECTION_NAME;
 pub use sdk_version_embedder::VERSION_SECTION_NAME;
 
+// logs will be printed only if debug feature is enabled
+#[cfg(feature = "debug")]
 #[allow(unused_variables)]
 pub(crate) fn log<S: AsRef<str>>(msg: S) {
-    // logs will be printed only if debug feature is enabled
-    #[cfg(feature = "debug")]
```
[the remark this answers is not in the document]
Comment (michaelvoronov) [reply crates/main/src/lib.rs:63]: Finally, It became
```
#[macro_export]
macro_rules! debug_log {
    ($msg_generator:expr) => {
        #[cfg(feature = "debug")]
        {
            let level = log::Level::Info as i32;
            let target = 0i32;
            let msg = $msg_generator;
            logger::log_utf8_string(level, target, msg.as_ptr() as i32, msg.len() as i32);
        }
    };
}

```

**Pull request merged**